### ISSUE TITLE
Saving with encoding_auto should use the encoding detected during loading for compatibility

### ISSUE THREAD

**Issue by JayXon** [zeux/pugixml#89]
The same should applies to BOM as well.
If the `xml_document` was not loaded from anywhere, it's created from scratch, then use UTF-8 as default.
Currently there's no way to know the encoding of the file other than checking it myself, but pugixml already did this when loading with `encoding_auto`, so it would be great if pugixml could save this information and make use of it.

**Comment by zeux** [comment]
This is a pretty significant behavior change. For some use cases it's a good change, but for a lot of cases it may not be - encoding_auto is the default value of the argument, so it'd change how doc.save_file() behaves.

It should be easy to add something like xml_document::encoding() that would return the encoding document was loaded from - then you can pass it to save_file/etc. Will that work for your use case?

**Comment by JayXon** [comment]
That will work too, thanks!

**Comment by zeux** [comment]
Oh. Wait. I'm starting to forget my own API - I just remembered that this feature... already exists! :)

xml_parse_result::encoding contains the source document encoding. So you can just remember it and pass it along to save/save_file.

**Comment by JayXon** [comment]
LOL, I didn't notice this either. Anyway, I just tested it and the encoding works.
How about BOM? I would like to preserve that too. Would you mind adding something like `bool has_bom;` in the xml_parse_result? If not I could also check it myself.

**Comment by zeux** [comment]
I realized I never replied to that :-/ My bad. I thought I did and explained the reasoning.

While it's pretty easy to add has_bom (may require tiny changes in the parser), changing xml_parse_result breaks ABI which I'm trying to maintain for 1.x versions (there was an accidental breakage or two but starting from I believe 1.4 through 1.7, and including current unreleased version, ABI should be the same) - this helps for some Linux distributions, or so I hear.

Anyway, I'll close this issue and open another one, just for BOM.
